signer/src/testing/context.rs

```diff
@@ -184,6 +248,327 @@ impl BitcoinInteract for WrappedMock<MockBitcoinInteract> {
+    settings: crate::config::Settings,
+    storage: Storage,
+    bitcoin: Bitcoin,
+    stacks: Stacks,
+}
+
+impl Default for ContextConfig<(), (), ()> {
+    fn default() -> Self {
+        Self {
+            settings: Settings::new_from_default_config().expect("failed to load default config"),
+            storage: (),
+            bitcoin: (),
+            stacks: (),
+        }
+    }
+}
+
+/// State for the builder pattern.
+pub trait BuilderState<Storage, Bitcoin, Stacks> {
+    /// Consumes the builder, returning its current internal configuration.
+    fn get_config(self) -> ContextConfig<Storage, Bitcoin, Stacks>;
+}
+
+/// A builder for creating a [`TestContext`].
+pub struct ContextBuilder<Storage, Bitcoin, Stacks> {
+    config: ContextConfig<Storage, Bitcoin, Stacks>,
+}
+
+impl ContextBuilder<(), (), ()> {
+    /// Create a new context builder.
+    pub fn new() -> Self {
+        Self { config: Default::default() }
+    }
+}
+
+impl Default for ContextBuilder<(), (), ()> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<Storage, Bitcoin, Stacks> BuilderState<Storage, Bitcoin, Stacks>
+    for ContextBuilder<Storage, Bitcoin, Stacks>
+{
+    fn get_config(self) -> ContextConfig<Storage, Bitcoin, Stacks> {
+        self.config
+    }
+}
+
+/// Trait for configuring the settings. These methods are always available.
+pub trait ConfigureSettings<Storage, Bitcoin, Stacks>
+where
+    Self: Sized + BuilderState<Storage, Bitcoin, Stacks>,
+{
+    /// Configure the context with the specified settings.
+    fn with_settings(self, settings: Settings) -> ContextBuilder<Storage, Bitcoin, Stacks> {
+        let config = self.get_config();
+        ContextBuilder {
+            config: ContextConfig {
+                settings,
+                storage: config.storage,
+                bitcoin: config.bitcoin,
+                stacks: config.stacks,
+            },
+        }
+    }
+}
+
+impl<Storage, Bitcoin, Stacks> ConfigureSettings<Storage, Bitcoin, Stacks>
+    for ContextBuilder<Storage, Bitcoin, Stacks>
+where
+    Self: BuilderState<Storage, Bitcoin, Stacks>,
+{
+}
+
+/// Trait for configuring the storage implementation. These methods are available
+/// when the storage implementation has not been set yet.
+pub trait ConfigureStorage<Bitcoin, Stacks>
+where
+    Self: Sized + BuilderState<(), Bitcoin, Stacks>,
+{
+    /// Configure the context with an in-memory storage implementation.
+    fn with_in_memory_storage(self) -> ContextBuilder<SharedStore, Bitcoin, Stacks> {
+        let config = self.get_config();
+        ContextBuilder {
+            config: ContextConfig {
+                settings: config.settings,
+                storage: Store::new_shared(),
+                bitcoin: config.bitcoin,
+                stacks: config.stacks,
+            },
+        }
+    }
+
+    /// Configure the context with the specified storage implementation.
+    fn with_storage<Storage: DbRead + DbWrite>(
```

Comment:
```suggestion
    fn with_storage<Storage: DbRead + DbWrite + Clone + Send + Sync>(
```

Reply:
updated in [f4951ef](https://github.com/stacks-network/sbtc/pull/606/commits/f4951ef29f8ea66b9a3c2ea436ff0d668395b9fa) :+1: